### PR TITLE
Correct assumptions for forwarded hostname & IP handling

### PR DESCRIPTION
**Pull request by tstromberg**
Follow-up to #7310 - addressing the hairier code review feedback.

- Docker on Linux should use a true IP/port when possible rather than forwarded IP/port
- `update-context` command should update the Docker port

Fixes an inconsistency:

- `start` saves --apiserver-name into kubeconfig, but `update-context` does not

Consistently fixing the last issue required replacing IP's with strings throughout.

## Work to be done

- Integration tests to ensure that things are done correctly.

### CODE REVIEW

File: test/integration/main.go [hunk 1]
```diff
@@ -73,6 +74,11 @@ func KicDriver() bool {
 	return strings.Contains(*startArgs, "--driver=docker") || strings.Contains(*startArgs, "--vm-driver=docker") || strings.Contains(*startArgs, "--vm-driver=podman") || strings.Contains(*startArgs, "driver=podman")
 }
 
+// NeedsPortForward requires whether or not this host needs port forwarding
```
Comment (medyagh) [inline test/integration/main.go:77]: coud maybe add more comments, for the contributors that dont know much about this.

docker on non-linux uses a bridge networking

File: test/integration/cert_options_test.go [hunk 2]
```diff
@@ -0,0 +1,57 @@
...
+	profile := UniqueProfileName("cert-options")
+	ctx, cancel := context.WithTimeout(context.Background(), Minutes(30))
+	defer CleanupWithLogs(t, profile, cancel)
+
+	// Use the most verbose logging for the simplest test. If it fails, something is very wrong.
+	args := append([]string{"start", "-p", profile, "--apiserver-ips=127.0.0.1,192.168.15.15", "--apiserver-names=localhost,www.google.com", "--apiserver-port=8555"}, StartArgs()...)
+
+	// We can safely override --apiserver-name with unique that works
+	if NeedsPortForward() {
+		args = append(args, "--apiserver-name=kubernetes.docker.internal")
```
Comment (medyagh) [inline test/integration/cert_options_test.go:42]: is this line okay with podman too ?

**Comment by medyagh** [comment]
this PR almost made my day ! thank you for doing this.

the concept looks good

**Comment by medyagh** [comment]
@tstromberg this test seems to be broken:

```
=== RUN   TestFunctional/parallel/ProfileCmd/profile_not_create
--- FAIL: TestFunctional/parallel/ProfileCmd/profile_not_create (10.81s)
functional_test.go:530: (dbg) Done: out/minikube-linux-amd64 profile list --output json: (10.703099107s)
functional_test.go:537: out/minikube-linux-amd64 profile list --output json failed: invalid character '*' looking for beginning of value
```

```

https://storage.googleapis.com/minikube-builds/logs/7360/ba84f94/Docker_Linux.html#fail_TestFunctional%2fparallel%2fProfileCmd%2fprofile_not_create

``

**Comment by tstromberg** [comment]
Thanks! I did some additional confirmation here locally on the Docker failures:

## TestStartStop/group/containerd - 215.51s
## TestStartStop/group/embed-certs - 173.21s

`Error from server (Forbidden): error when creating "testdata/busybox.yaml": pods "busybox" is forbidden: error looking up service account default/default: serviceaccount "default" not found`

* **Known flake**: https://github.com/kubernetes/minikube/issues/7011

## TestStartStop/group/crio - 782.71s

`Warning  FailedCreatePodSandBox  7m18s                 kubelet, crio-20200401t175453.296999333-32003  Failed create pod sandbox: rpc error: code = Unknown desc = failed to create pod network sandbox k8s_busybox_default_1aa75488-2899-4d7d-98cb-50dc9ff62636_0(b2c34f2a8500f6f48f8d267786cbf3a04834cc756f94052f391ef074f5aac7a4): failed to set bridge addr: could not add IP address to "cni0": permission denied`

* **Recent unrelated regression**: https://github.com/kubernetes/minikube/issues/7380
* **Reproducible on macOS**: Yes
* **Reproducible on Linux**:  Yes

## TestFunctional/parallel/ServiceCmd - 69.9s

`functional_test.go:648: (dbg) Run:  out/minikube-linux-amd64 -p functional-20200401T175116.759279844-32003 service hello-node --url
functional_test.go:662: url: http://172.17.0.2:32093
functional_test.go:665: get failed: GET http://172.17.0.2:32093 giving up after 5 attempts`
 

* *Reproducible on macOS**: No - but did open https://github.com/kubernetes/minikube/issues/7383
* **Reproducible on Linux**: No
* **NOTE**: This seems to be a very common flake looking at the Jenkins log

## TestFunctional/parallel/ProfileCmd/profile_not_create - 10.81s

`functional_test.go:537: out/minikube-linux-amd64 profile list --output json failed: invalid character '*' looking for beginning of value`

* **Opened issue**: https://github.com/kubernetes/minikube/issues/7381
* **PR**: https://github.com/kubernetes/minikube/pull/7382/files

**Comment by medyagh** [comment]
Thank you for creating the issues.i recommend putting this comment of the PR as a required model of response for all future PRs.

Maybe add it to the contributor guide?( In a separate pr)

How to handle integration tests failure .

looks good we can merge

File: test/integration/functional_test.go [hunk 3]
```diff
@@ -335,8 +335,9 @@ func validateDashboardCmd(ctx context.Context, t *testing.T, profile string) {
 
 	resp, err := retryablehttp.Get(u.String())
 	if err != nil {
-		t.Fatalf("failed to http get %q : %v", u.String(), err)
```
Comment (medyagh) [inline test/integration/functional_test.go:338]: can we have the URL back in to the 
failed to http get ... ?

Comment (tstromberg) [reply test/integration/functional_test.go:338]: It's still there.

**Pull request merged**